op-node/rollup/clsync/clsync.go

```diff
@@ -0,0 +1,130 @@
+package clsync
+
+import (
+	"context"
+	"errors"
+	"io"
+
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum-optimism/optimism/op-node/rollup/derive"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+)
+
+// Max memory used for buffering unsafe payloads
+const maxUnsafePayloadsMemory = 500 * 1024 * 1024
+
+type Metrics interface {
+	RecordUnsafePayloadsBuffer(length uint64, memSize uint64, next eth.BlockID)
+}
+
+type Engine interface {
+	derive.EngineState
+	InsertUnsafePayload(ctx context.Context, payload *eth.ExecutionPayloadEnvelope, ref eth.L2BlockRef) error
+}
+
+// CLSync holds on to a queue of received unsafe payloads,
+// and tries to apply them to the tip of the chain when requested to.
+type CLSync struct {
+	log            log.Logger
+	cfg            *rollup.Config
+	metrics        Metrics
+	ec             Engine
+	unsafePayloads *PayloadsQueue // queue of unsafe payloads, ordered by ascending block number, may have gaps and duplicates
+}
+
+func NewCLSync(log log.Logger, cfg *rollup.Config, metrics Metrics, ec Engine) *CLSync {
```

Comment:
I think it'd be nice to find a different name for this package / struct. Medium term I think we should use the unsafe payload queue for EL sync. Specifically once you do the first newPayload/FCU call to geth, it prefers having the blocks inserted in order (particularly when doing a full sync).

Reply:
Currently it's separate still, EL sync completely bypasses it. When EL sync does utilize it with a later change we can always rename it.